File: lab3/src/geometry/Point.java
package geometry;

public class Point {
    
    private double x;
    private double y;

    public Point(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public double getX() {
        return this.x;
    }

    public double getY() {
        return this.y;
    }

    public void setX(double value) {
        this.x = value;
    }

    public void setY(double value) {
        this.y = value;
    }

    public void move(double offsetX, double offsetY) {
        this.x += offsetX;
        this.y += offsetY;
    }

    public double magnitude() {
        return Math.sqrt(
            x * x + y * y
        );
    }

    @Override
    public String toString() {
        return String.format("[%.2f, %.2f]", this.x, this.y);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Point) {
            var point = (Point)obj;
            return this.x == point.x && this.y == point.y;
        }
        
        return false;
    }

    public static double dot(Point v1, Point v2) {
        return v1.x * v2.x + v1.y * v2.y;
    }

    public static Point copy(Point other) {
        return new Point(other.x, other.y);
    }
}

File: lab5/src/geometry/Point.kt
package geometry

import kotlin.math.sqrt

class Point(private var x: Double, private var y: Double) {

    fun getX(): Double {
        return this.x
    }

    fun getY(): Double {
        return this.y
    }

    fun setX(value: Double) {
        this.x = value
    }

    fun setY(value: Double) {
        this.y = value
    }

    fun move(offsetX: Double, offsetY: Double) {
        this.x += offsetX
        this.y += offsetY
    }

    fun magnitude(): Double {
        return sqrt(
            x * x + y * y
        )
    }

    override fun toString(): String {
        return String.format("[%.2f, %.2f]", this.x, this.y)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Point) return false

        return this.x == other.x &&
                this.y == other.y
    }

    override fun hashCode(): Int {
        var result = x.hashCode()
        result = 31 * result + y.hashCode()
        return result
    }

    companion object {
        fun dot(v1: Point, v2: Point): Double {
            return v1.x * v2.x + v1.y * v2.y
        }

        fun copy(other: Point): Point {
            return Point(other.x, other.y)
        }
    }
}
